### PR TITLE
Add model merging utilities and tests for multi-model optimization

### PR DESCRIPTION
**Pull request by cetagostini**
## Description

Introduces functions and a wrapper class in `pytensor_utils.py` to merge multiple PyMC models with variable prefixing and shared variable support, enabling multi-model budget optimization. Updates `budget_optimizer.py` to use delayed imports to avoid circular dependencies. Adds tests for model merging and the new wrapper in `test_pytensor_utils.py`.
Combines multiple independent PyMC Models into a single Model.

### Example
As we move forward in causality, and discover new DAGs, a way to collect direct or indirect effects could be create different regressions, every one adjusting by the minimum set of variables. _Each regression will respond to the same DAG but will uncover different effects per channel depending on the adjustment_.

Nevertheless, we probably want to optimize all regressions together or we want to join the mode to make several operations. But how to? Well let's remember PyMC models are graphical generative model so, each model can be represent as a function graph, and their nodes/variables could be prefixed (e.g., model1_, model2_). Doing so, we could decide which variable or data will be the one to merge_on="some_var" and the variable is kept **unprefixed** and shared across the merged models. All outputs and coordinates are merged into a single function graph.

Imagine you have the following DAG:
```
digraph G {
  rankdir=LR; splines=true; nodesep=0.4; ranksep=0.5;
  node [shape=ellipse, fontsize=12];

  C1 -> X1; C1 -> Y;
  C2 -> X2; C2 -> Y;
  C3 -> X3; C3 -> Y;

  X1 -> X2; X2 -> X3;
  X1 -> Y;  X2 -> Y;  X3 -> Y;

  // styling
  {rank=same; C1; C2; C3}
  {rank=same; X1; X2; X3}
  Y [shape=doublecircle];
}
```

This structure makes the minimal sufficient sets different for each channel’s **direct** effect on Y:

- **For x1:** backdoor via C1; mediated paths via X2 to X3.  
  **Adjust:** C1,X2 (conditioning on X2 also blocks the X1 to X2 to X3 to Y path).

- **For X2:** backdoors via C2 and via X1 (since X1 to X2 and X1 to Y); mediated path via X3.  
  **Adjust:** C2, X1, X3.

- **For X3:** backdoors via C3 and via X2 (since X2 to X3 and X2 to Y).  
  **Adjust:** C_3, X_2.

Meaning, we need three models to uncover the effects:
```python
# Get x1 over Y
mask = np.array([True, False, False])
mmm_x1_effect = MMM(  # (Y ~ X1 + C1 + X2)
    date_column="date",
    target_column="Y",
    channel_columns=[
        "X1",
        "X2",
        "X3",
    ],  # always use all channels, but mask the ones that are not used
    control_columns=["X2", "C1"],
    adstock=NoAdstock(l_max=1),
    saturation=NoSaturation(
        priors={
            "beta": MaskedDist(
                Prior("HalfNormal", sigma=1.0, dims="channel"), mask=mask
            )
        }
    ),
)

# get x2 over Y
# make mask for channels making x1,x2 false
mask2 = np.array([False, True, False])
mmm_x2_effect = MMM(  # (Y ~ X2 + C2 + X1 + X3)
    date_column="date",
    target_column="Y",
    channel_columns=[
        "X1",
        "X2",
        "X3",
    ],  # always use all channels, but mask the ones that are not used
    control_columns=["X1", "C2", "X3"],
    adstock=NoAdstock(l_max=1),
    saturation=NoSaturation(
        priors={
            "beta": MaskedDist(
                Prior("HalfNormal", sigma=1.0, dims="channel"), mask=mask2
            )
        }
    ),
)

# get x3 over Y
# make mask for channels making x1,x2 false
mask3 = np.array([False, False, True])
mmm_x3_effect = MMM(  # (Y ~ X3 + C3 + X2)
    date_column="date",
    target_column="Y",
    channel_columns=[
        "X1",
        "X2",
        "X3",
    ],  # always use all channels, but mask the ones that are not used
    control_columns=["X2", "C3"],
    adstock=NoAdstock(l_max=1),
    saturation=NoSaturation(
        priors={
            "beta": MaskedDist(
                Prior("HalfNormal", sigma=1.0, dims="channel"), mask=mask3
            )
        }
    ),
)
```

As you can see we should be able to _zero out with mask the distributions to avoid calculate the contributions for the channels we don't want to adjust in the regression_. Nevertheless, even when channels are zero out given the priors and not consider during the sampling, the data input is the same for all.

We could merge based on it.

```python
optimizable_model1 = MultiDimensionalBudgetOptimizerWrapper(
    model=mmm_x1_effect, start_date="2026-01-01", end_date="2026-01-31"
)

optimizable_model2 = MultiDimensionalBudgetOptimizerWrapper(
    model=mmm_x2_effect, start_date="2026-01-01", end_date="2026-01-31"
)

optimizable_model3 = MultiDimensionalBudgetOptimizerWrapper(
    model=mmm_x3_effect, start_date="2026-01-01", end_date="2026-01-31"
)

multi_model = BuildMergedModel(
    models=[optimizable_model1, optimizable_model2, optimizable_model3],
    prefixes=["model1", "model2", "model3"],
    merge_on="channel_data",
)
```

And finally, optimize.

```python
optimizer = BudgetOptimizer(
    num_periods=multi_model.num_periods,
    model=multi_model,
    response_variable="total_media_contribution_original_scale",
)
```

This opens the door to make optimization of different effects (total, indirects or directs) together. Allows for multi-objective optimization. Every model could have a different target variable but the target depends on same inputs, and consequently we want to optimize A and B together.

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

### CODE REVIEW

**Comment by juanitorduz** [comment]
Thanks @cetagostini ! Im looking into the codecov report, and it seems we are still not testing all of the edge cases :) Would you mind adding tests for these ones please :)  Thanks!

File: pymc_marketing/pytensor_utils.py [hunk 1]
```diff
@@ -14,20 +14,326 @@
...
+    # Merge coordinates from all models
+    merged_coords: dict = {}
+    for fg in fgraphs:
+        merged_coords.update(fg._coords)  # type: ignore[attr-defined]
+    f._coords = merged_coords  # type: ignore[attr-defined]
+
+    return model_from_fgraph(f, mutate_fgraph=True)
+
+
+class BuildMergedModel(OptimizerCompatibleModelWrapper):
```
Comment (williambdean) [inline pymc_marketing/pytensor_utils.py]: If this is causing circular imports, it should be somewhere else. Why not in the budget optimizer module since it is specific to that?

Comment (cetagostini) [reply pymc_marketing/pytensor_utils.py]: I was thinking it could be use for more than optimizer, you could actually merge models and them sample the full merge model, nevertheless, make sense to have the class there. I'll be moving!

**Comment by cetagostini** [comment]
@juanitorduz or @williambdean ready for review!

**Review by juanitorduz — APPROVED** [review]
Thanks @cetagostini ! Do you plan to add a example notebook showcasing this ?
